File: Chapter 10/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter10/section_10_3_resilience_decorators/resilient/infrastructure/adapters/FlakyPaymentsPaymentAdapter.java
package com.grokkingsoftwarearchitecture.chapter10.section_10_3_resilience_decorators.resilient.infrastructure.adapters;

import com.grokkingsoftwarearchitecture.chapter10.section_10_3_resilience_decorators.resilient.core.ports.PaymentGateway;
import io.github.resilience4j.core.IntervalFunction;
import io.github.resilience4j.retry.Retry;
import io.github.resilience4j.retry.RetryConfig;
import io.github.resilience4j.core.functions.CheckedSupplier;
import java.time.Duration;

/**
 * THE INFRASTRUCTURE ADAPTER (The Implementation):
 * * @description
 * This class encapsulates the Physical Resource Policy for FlakyPayments using Resilience4j. 
 *
 * ARCHITECTURAL CRITIQUE:
 * 1. OBSERVABILITY: We use a custom 'onRetry' event listener to match our 
 * high-rigor JS/C# logging. This ensures every attempt and backoff is 
 * visible in the lab trace.
 * * 2. FAIL-FAST: By throwing the exception immediately on the final attempt, 
 * we prevent "Temporal Leakage" where the system waits for a useless 
 * backoff period before failing over to Plan B.
 */
public class FlakyPaymentsPaymentAdapter implements PaymentGateway {

    // --- THE PHYSICAL POLICY CONSTANTS (The SLA) ---
    private static final int MAX_RETRIES = 5;
    private static final int MAX_ATTEMPTS = MAX_RETRIES + 1;
    private static final long MIN_DELAY_MS = 2000;
    private static final long MAX_DELAY_MS = 10000;
    private static final double BACKOFF_FACTOR = 2.0;

    private final String baseUrl;
    private final Retry retryShield;

    public FlakyPaymentsPaymentAdapter(String baseUrl) {
        this.baseUrl = baseUrl;

        // THE SHIELD (Declarative Policy via Resilience4j)
        RetryConfig config = RetryConfig.custom()
                .maxAttempts(MAX_ATTEMPTS)
                .intervalFunction(IntervalFunction.ofExponentialBackoff(
                        MIN_DELAY_MS, 
                        BACKOFF_FACTOR, 
                        MAX_DELAY_MS))
                .retryExceptions(Exception.class)
                .build();

        this.retryShield = Retry.of("FlakyPayments-charge-retry", config);

        // LOGGING SHIELD: Synchronizing with the trace
    this.retryShield.getEventPublisher().onRetry(event -> {
            int retryAttempt = event.getNumberOfRetryAttempts();
            System.out.println("      [Retry Shield] Attempt " + retryAttempt + " failed: " + event.getLastThrowable().getMessage() + ".");
            
            // If we have finished our last allowed retry (Attempt 5)
            if (retryAttempt >= MAX_RETRIES) {
                System.err.println("      [Retry Shield] MAX_RETRIES (" + MAX_RETRIES + ") reached. Exhausted.");
            } else {
                System.out.println("      [Retry Shield] Backing off " + event.getWaitInterval().toMillis() + "ms...");
            }
        });
    }

    @Override
    public boolean charge(double amount, String orderId, String idempotencyKey) {
        // FIX: Explicitly type the CheckedSupplier to Boolean to solve the incompatible types error
        CheckedSupplier<Boolean> resilientCall = Retry.decorateCheckedSupplier(retryShield, () -> {
            System.out.println("      [FlakyPayments Adapter] Attempting FlakyPayments Charge for " + orderId + "...");
            
            // SIMULATION: Triggering the shield as seen in your other implementations
            throw new RuntimeException("FlakyPayments API: Gateway Timeout (504)");
        });

        // Convert the checked supplier into an unchecked execution
        try {
            return resilientCall.get();
        } catch (Throwable throwable) {
            // Re-throw as RuntimeException to allow Orchestrator catch block to handle it
            throw new RuntimeException(throwable.getMessage(), throwable);
        }
    }
}

File: Chapter 10/C#/Section_10_3_Resilience_Decorators/Resilient/Infrastructure/FlakyPaymentsPaymentAdapter.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;
using Polly;
using Polly.Retry;
using Chapter10.Resilient.Core.Ports;

namespace Chapter10.Resilient.Infrastructure.Adapters;

/// <summary>
/// THE INFRASTRUCTURE ADAPTER (The Implementation):
///
/// This class encapsulates the Physical Resource Policy for the FlakyPayments vendor.
/// It uses the Polly library to implement a declarative "Retry Shield."
///
/// ARCHITECTURAL CRITIQUE:
/// 1. OBSERVABILITY: Following the JavaScript standard, we now log every 
/// failure and the specific backoff duration. This prevents "Silent Failures" 
/// and makes the system's struggle visible to the operator.
/// 2. FAIL-FAST: By using the 'onRetry' hook, we can detect when we are 
/// about to hit the limit and ensure the final failure is explicitly marked 
/// as 'Exhausted' before the Orchestrator takes over.
/// 3. Physical Resource Policy for the FlakyPayments vendor.
/// By moving our SLA (Service Level Agreement) into named constants, we 
/// transform hidden magic numbers into a documented, tunable boundary.
/// The Core Application remains pure because the Polly retry policy is 
/// physically locked inside this adapter.
/// </summary>
public class FlakyPaymentsPaymentAdapter : IPaymentGateway
{
    private const int TotalRequestTimeoutSec = 10;
    private const int MaxRetryAttempts = 5;
    private const double InitialDelayMs = 2000;
    private const double MaxDelayMs = 10000;
    private const int BackoffFactor = 2;

    private readonly HttpClient _httpClient;
    private readonly AsyncRetryPolicy<bool> _retryPolicy;

    public FlakyPaymentsPaymentAdapter(string baseUrl)
    {
        _httpClient = new HttpClient 
        { 
            BaseAddress = new Uri(baseUrl),
            Timeout = TimeSpan.FromSeconds(TotalRequestTimeoutSec)
        };

        // THE SHIELD (Declarative Policy)
        _retryPolicy = Policy<bool>
            .Handle<HttpRequestException>()
            .Or<TaskCanceledException>()
            .WaitAndRetryAsync(
                retryCount: MaxRetryAttempts,
                sleepDurationProvider: retryAttempt => {
                    // Exponential Backoff calculation
                    var backoff = InitialDelayMs * Math.Pow(BackoffFactor, retryAttempt - 1);
                    return TimeSpan.FromMilliseconds(Math.Min(backoff, MaxDelayMs));
                },
                onRetryAsync: async (outcome, timespan, retryCount, context) => {
                    // LOG THE FAILURE
                    Console.WriteLine($"      [Retry Shield] Attempt {retryCount} failed: {outcome.Exception?.Message ?? "Timeout"}.");

                    // If this is our last permitted retry, log the exhaustion
                    if (retryCount >= MaxRetryAttempts)
                    {
                        Console.WriteLine($"      [Retry Shield] MAX_RETRIES ({MaxRetryAttempts}) reached. Exhausted.");
                    }
                    else
                    {
                        Console.WriteLine($"      [Retry Shield] Backing off {timespan.TotalMilliseconds}ms...");
                    }
                    
                    await Task.CompletedTask;
                }
            );
    }

    public async Task<bool> ChargeAsync(decimal amount, string orderId, string idempotencyKey)
    {
        return await _retryPolicy.ExecuteAsync(async () =>
        {
            Console.WriteLine($"      [FlakyPayments Adapter] Attempting FlakyPayments Charge for {orderId}...");

            _httpClient.DefaultRequestHeaders.Clear();
            _httpClient.DefaultRequestHeaders.Add("Idempotency-Key", idempotencyKey);

            // SIMULATION: Throwing an error to trigger the shield as seen in your JS example
            throw new HttpRequestException("FlakyPayments API: Gateway Timeout (504)");
            
            /* // Real implementation would look like this:
            var response = await _httpClient.PostAsJsonAsync("/charge", new { amount, order_id = orderId });
            response.EnsureSuccessStatusCode();
            return true;
            */
        });
    }
}